### ISSUE TITLE
remains after kakoune exits

### ISSUE THREAD

**Issue by nrdxp** [kak-lsp/kak-lsp#518]
On latest release, kak-lsp process seems to linger even after kakoune process has exited. I frequently have to kill the process with `pkill`.

**Comment by nrdxp** [comment]
nvm it was a timeout config issue. apologies

**Comment by krobelus** [comment]
Our `KakEnd` hook should stop `kak-lsp` when a Kakoune session exits, independent of the timeout in `kak-lsp.toml`

**Comment by nrdxp** [comment]
Ugh, I'm very sorry for this, but I'm gonna have to reopen actually as I tried removing the timeout to see if it would fix, and it didn't happen about 5 times in a row so I thought it did, but I just ran:
```
❯ pgrep kak-lsp -a
352323 /nix/store/9fby35g2f5zan3vbmsmm60wxvnkb2j7j-kak-lsp-11.0.0/bin/kak-lsp -s 348420 --daemonize --initial-request
354653 /nix/store/9fby35g2f5zan3vbmsmm60wxvnkb2j7j-kak-lsp-11.0.0/bin/kak-lsp -s 353798 --daemonize --initial-request
360433 /nix/store/9fby35g2f5zan3vbmsmm60wxvnkb2j7j-kak-lsp-11.0.0/bin/kak-lsp -s 352341 --daemonize --initial-request
```
after an expended editing session and the issue still appears to be happening. I did not have a running kakoune instance at the time of this post.

**Comment by krobelus** [comment]
Is this with the default configuration (for Kakoune and kak-lsp)? Is the language server relevant? Does it happen every time? Does it work if you try the same with the previous release?

Here's a quick way to test the default configuration

```
HOME=$(mktemp -d) kak -e '
	eval %sh{kak-lsp --kakoune -s $kak_session}
	set global lsp_cmd "kak-lsp -s %val{session} -vvvv --log /tmp/kak-lsp.log"
	lsp-enable'
```

**Comment by nrdxp** [comment]
I just finished pulling in the default config toml with only a single change (using rust-analyzer instead of rls). I haven't tried with any other lsp just yet.

**Comment by stacyharper** [comment]
I also am encountering this. It seems to happen randomly, and I can't reproduce this reliably. Maybe it is caused by a specific workflow, or a backend that hang indefinitely

**Comment by Jackson-Nickk** [comment]
I encounter this when using a bash language server. It happens with bot `bash-language-server` and `bashd`.

There's also this weird behavior: this problem doesn't happen when editing my old scripts that were created before installing and configuring kak-lsp. It only happens when editing my newer scripts.

**Comment by Jackson-Nickk** [comment]
Now it also happens with marksman!

**Comment by krobelus** [comment]
> I encounter this when using a bash language server. It happens with bot bash-language-server and bashd.

does the same happen with at least one other LSP client like Neovim, Helix or VSCode?
If yes, then it's likely a server issue.

If no, then let's try to carve out a portable reproducer

**Comment by krobelus** [comment]
should be fixed with #892

**Comment by Jackson-Nickk** [comment]
> > I encounter this when using a bash language server. It happens with bot bash-language-server and bashd.
> 
> does the same happen with at least one other LSP client like Neovim, Helix or VSCode? If yes, then it's likely a server issue.
> 
> If no, then let's try to carve out a portable reproducer

It's kak-lsp that remains running after kakoune is closed. kak-lsp is not available for other editors. Helix closes its LSP client gracefully as far as I can tell.

**Comment by krobelus** [comment]
Do the language servers keep running as well?
Let's find a reproducer, or, when it happens, attach `rust-gdb` to the stuck kak-lsp process (I'm assuming kak has already exited),
and check where the threads are hanging.
As mentioned before, the `KakEnd` hook should send `kakoune/exit` which should make kak-lsp exit, after sending `exit` to language servers..

**Comment by Jackson-Nickk** [comment]
Yes. The language server (`bashd`) also keeps running. Before `bashd` I was using the `bash-language-server` and this situation was present. I switched to `bashd` because I thought the minimalism of Golang would solve the slowness of JavaScript, which is what `bash-language-server` is written in.

I have a hook that defines `bashd` as my language server for bash and sh:

```kakscript
remove-hooks global lsp-filetype-sh
 hook -group lsp-filetype-sh global BufSetOption filetype=(sh|bash) %{
   set-option buffer lsp_servers %{
     [bashd]
     root_globs = [".git"]
     command = "bashd"
   }
 }
```
When I enable this hook, and open a newly created shell script, then try to exit kakoune, kakoune hangs and pgrep tells me that both `kak <script>` and `kak-lsp --daemonize` processes are running. Killing `bashd` causes both kak and kak-lsp to exit out of their hang.

This hang doesn't happen with `clangd` and `marksman`.

Attaching rust-gdb to the hanging kak-lsp process using th `-p` option with root privileges gives me this:

```log
Attaching to process 19098
[New LWP 19132]
[New LWP 19131]
[New LWP 19130]
[New LWP 19104]
[New LWP 19103]
[New LWP 19101]
[New LWP 19099]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib64/libthread_db.so.1".
syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38

warning: 38	../sysdeps/unix/sysv/linux/x86_64/syscall.S: No such file or directory
```

I'm not familiar with GDB. This is my first time using it. But typing `help` in GDB's prompt gave me the idea of using the `trace` command, and the output of `trace` is:
```log
Tracepoint 1 at 0x7fa204765789: file ../sysdeps/unix/sysv/linux/x86_64/syscall.S, line 38.
```

Sorry if I'm not competent enough for this procedure. Tell me any other testing you might want me to do.

**Comment by krobelus** [comment]
@Jackson-Nickk I can't reproduce; again if this triggers on initially-empty files, this is very likely #892, fixed in kak-lsp `v21.0.1`. So please make sure to test with that or master.
To elaborate, I can't reproduce with `bash-language-server`, and I don't know where to find `bashd`.

If it still reproduces, give precise and isolated repro instructions, i.e. using `HOME=$(mktemp -d)` etc. similar to what we have in the README.

In GDB, you'd need to look at `info threads` and then find the thread that's stuck but shouldn't be.. likely no need to do this.

**Comment by Jackson-Nickk** [comment]
@krobelus 
`v21.0.1` hasn't reached the VoidLinux repo yet. I installed [bashd](https://github.com/matkrin/bashd) from Void's repo too. 

I'll wait for the new version to reach Void repo and test it.
